File: leetcode/linked_list/easy/golang_solutions/linked_list_cycle_141.go
package golang_solutions

import "github.com/vielendanke/algorithms_in_place/leetcode/linked_list"

func detectCycle(head *linked_list.ListNode) *linked_list.ListNode {
	visited := make(map[*linked_list.ListNode]interface{})

	for head != nil {
		if _, ok := visited[head]; ok {
			return head
		}
		visited[head] = nil
		head = head.Next
	}
	return head
}

func detectCycleWithoutMap(head *linked_list.ListNode) *linked_list.ListNode {
	slow := head
	fast := head

	if slow == nil && fast == nil {
		return head
	}
	for fast != nil && fast.Next != nil {
		slow = slow.Next
		fast = fast.Next.Next
		if slow == fast {
			return slow
		}
	}
	return nil
}

File: leetcode/linked_list/easy/java_solutions/LinkedListCycle_141.java
package leetcode.linked_list.easy.java_solutions;

import java.util.HashSet;
import java.util.Set;

import static leetcode.linked_list.Helper.ListNode;

public class LinkedListCycle_141 {

    static class Solution {
        public boolean hasCycle(ListNode head) {
            if (head == null) return false;
            ListNode slow = head;
            ListNode fast = head;
            while (fast != null && fast.next != null) {
                slow = slow.next;
                fast = fast.next.next;
                if (slow == fast) return true;
            }
            return false;
        }
    }

    static class SolutionBruteForce {
        public boolean hasCycle(ListNode head) {
            Set<ListNode> set = new HashSet<>();

            while (head != null) {
                if (set.contains(head)) {
                    return true;
                }
                set.add(head);
                head = head.next;
            }
            return false;
        }
    }
}
